File: src/09.tree/48.669.go
package tree

func TrimBST(root *TreeNode, low int, high int) *TreeNode {
	if root == nil {
		return nil
	}

	if root.Val < low {
		// 左子树不符合, 右子树可能符合
		return TrimBST(root.Right, low, high)
	}

	if root.Val > high {
		// 右子树不符合, 左子树可能符合
		return TrimBST(root.Left, low, high)
	}

	root.Left = TrimBST(root.Left, low, high)
	root.Right = TrimBST(root.Right, low, high)

	return root
}

File: src/09.tree/48.669.ts
import { TreeNode } from "./01.100";

export function trimBST(
    root: TreeNode | null,
    low: number,
    high: number
): TreeNode | null {
    if (!root) {
        return null;
    }

    // 不符合时, 总会返回正确的 root
    if (root.val < low) {
        return trimBST(root.right, low, high);
    } else if (root.val > high) {
        return trimBST(root.left, low, high);
    }

    // 符合时, 递归处理子树
    root.left = trimBST(root.left, low, high);
    root.right = trimBST(root.right, low, high);

    return root;
}
